### PR TITLE
fix(client): report the debt on single-cosmetic purchase (OPE-373)

### PR DESCRIPTION
**Pull request by Celant**
## What

`purchaseWithCurrency` (`POST /shop/purchase`, every single-cosmetic buy, and the only spend path that also takes soft currency) returned a bare boolean and never read the response body. It now returns a result union with `insufficient_balance` and `debt` variants, mirroring the pack path, and the caller explains a debt instead of showing a generic "Purchase failed".

## Why

`insufficientBalance()` in the API refuses with `{reason: "insufficient_balance_debt", debt}` when a refund or chargeback has left the wallet negative. This path discarded the body entirely, so that was indistinguishable from a timeout. Steam adds a second chargeback rail with a much shorter fraud window, so the population that can hit it grows on day one.

### The part that makes this actually reachable

`/users/@me` serves a charged-back wallet as a **negative** `currency.hard` — the API derives its own `debt` field as exactly `-hard`. So the ordinary way a debtor arrives here is not the server refusal at all: they fail the *client-side* balance pre-check and get the top-up dialog ("you need 400 more") with a button that cannot clear a debt, without a request ever being made. The result-code branch alone would only have fired when a chargeback landed mid-session against a stale cache.

Both pre-checks (single cosmetic and pack) now test for a negative balance first. Without this the ticket's goal is essentially unreachable in production.

Three smaller consequences, all on the same theme of not telling the player something untrue:

- If the post-refusal re-read still covers the price, there is no shortfall to quote — it reports a generic failure rather than inventing "you need 1 more" for a purchase that would now succeed.
- If the post-refusal re-read comes back negative, it explains the debt.
- Both paths now `invalidateUserMe()` on a debt refusal. The pack path had the same omission; the store was left rendering the pre-chargeback balance.

A debt amount that is not a run of digits falls back to a generic failure rather than rendering a blank, or `[object Object]`, where the amount belongs. The amount is the whole message, so a debt we cannot state is worse than no debt message.

Reasons are allowlisted, not denylisted: anything unrecognised is a generic failure, with a `console.warn` that deliberately omits the body.

## Tests

- `tests/client/CosmeticPurchaseDebt.test.ts` — the caller's branches, mirroring `CosmeticPackPurchase.test.ts`.
- New `purchaseWithCurrency` describe in `tests/client/PurchaseCosmeticPack.test.ts` — the reason mapping over a mocked fetch, including a table of unusable debt amounts.
- Negative-balance pre-check coverage on both the single-cosmetic and pack paths, asserting no request is made at all.

**19 of the 30 fail against `origin/main`.**

```
npx vitest tests/client/CosmeticPurchaseDebt.test.ts tests/client/PurchaseCosmeticPack.test.ts \
  tests/client/CosmeticPackPurchase.test.ts --run                       # 30 passed
npx vitest <those + CosmeticsPaymentsMigration, PurchaseReturn,
  CosmeticLockerIntegration, PurchaseButton, StoreModal, InventoryModal> # 124 passed
npx tsc --noEmit                                                        # clean
npm run lint                                                            # clean
```

## Notes

- One test was deleted rather than fixed: it asserted the machine reason never reached the message while the mock itself supplied `code: "debt"`, so the string existed nowhere in the fixture and it could not fail. Replaced by assertions that an unrecognised 400 carries no reason back to the caller and that the warn takes exactly one argument.
- The equivalent `debt`-validation tightening is **not** applied to the tribe-name path on #5307, which is left at parity with the pre-existing pack behaviour. If #5307 is re-pushed for another reason it should come along.
- Out of scope, filed separately: this path folds the API's 409 "Already owned" into a generic failure, so a retry after a timed-out success cannot converge (the pack path maps 409 to `already_owned` + reload).

Resolves OPE-373 — https://linear.app/openfront/issue/OPE-373

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/client/Cosmetics.ts [hunk 1]
```diff
@@ -461,15 +481,57 @@ export async function purchaseCosmetic(
...
+        // The balance moved since the pre-check: re-read it for the shortfall,
+        // the same way the pack path does.
+        invalidateUserMe();
+        const fresh = await getUserMe();
+        const available =
+          fresh === false
+            ? 0
+            : method === "hard"
+              ? (fresh.player.currency?.hard ?? 0)
+              : (fresh.player.currency?.soft ?? 0);
```
[the remark this answers is not in the document]
Comment (Celant) [reply src/client/Cosmetics.ts]: Agreed and fixed in d285e97.

You are right: falling back to `0` made the shortfall the full price, which is not a measurement — the re-read failed, so the real balance is unknown. It then rendered with a top-up button sized to that guess.

A failed re-read is now treated as an unexplained failure (`store.purchase_failed`) rather than a shortfall, on both the single-cosmetic and pack paths.

While in there, the same branch had two neighbouring cases with the same flavour of problem, so all three now go through one shared classifier:

- re-read is **negative** → explain the debt (a chargeback landed since the pre-check; topping up cannot clear it)
- re-read **covers the price** → generic failure, rather than the clamp's fabricated "you need 1 more" for a purchase that would now succeed
- re-read is genuinely **short** → the shortfall dialog, as before

The clamp survives only as a floor. Tests cover each outcome on both paths, plus a zero-balance boundary case per pre-check so that an empty wallet is never mistaken for a debt.

The test you would have been pointing at, "still quotes a shortfall when the refetch fails", is inverted accordingly — it now asserts no dialog and a generic failure. The equivalent change is applied on #5307 so the two PRs agree.

**Review by evanpelle — APPROVED** [review]
